### ISSUE TITLE
Provide a report allowing a user to review their imports

### ISSUE THREAD

**Issue by johnvanbreda** [BiologicalRecordsCentre/iRecord#1294]
If an import fails for any reason, it can be difficult to discern which records were imported before the failure. To help in these situations, provide a report that lists a user's recent imports with information on the first and last record IDs included in the report.

**Comment by kitenetter** [comment]
Does the importer always import records in the same order as the order of the rows in the source file? I.e. if the report tells me that the latest record imported is record X, can I be certain that all rows in my source file prior to record X have been imported?

**Comment by johnvanbreda** [comment]
@kitenetter yes, apart from rows that had errors.

This assumption may not be true for v2, as it may sort the file into sample order for example.

**Comment by DavidHepper** [comment]
Please add a footnote to the form indicating that records imported in Training Mode can only be opened from the links while in Training Mode.

**Comment by kitenetter** [comment]
Need to clarify how this relates to the new importer - to be tested as part of #1405

**Comment by cubitmg** [comment]
I think there should be a way to list all records from an individual  import, presumably using a filter. I think the import date/time is set in the "Submission date" / "Entered on" date and can be filtered on using "Input date", but if more than one import is done on a date then this doesn't 'work'. 
I also think that there should be a way to delete all records from an import i.e. cancel the import. This could either be because there has been a technical issue importing or there may be a data problem with the imported records.

**Comment by DavidHepper** [comment]
@cubitmg I've just been asked to rollback an import and support your request, though not at the highest priority.

**Comment by DavidHepper** [comment]
It's really good to have the new view! 
I suggest it would be useful to have info at the top to inform users that records are not necessarily contiguous between the From and To IDs.

**Comment by kitenetter** [comment]
Review and complete - see also #1615

**Comment by DavidHepper** [comment]
The report could usefully include a flag indicating whether the records were imported in Training Mode. Otherwise clicking on one of the record numbers shows 'The record cannot be found'. Alternatively and usefully, this error message could be changed to give some options for why it can't be found.

**Comment by cubitmg** [comment]
Very useful to have the page <https://irecord.org.uk/imports/my-list> with the list of imports. Thank you.

Rather than an indicator about whether an import was carried out in Training Mode, I think that the imports/my-list should show only the imports for the mode that the user is in. So, in my case I would see 12 imports when in Training Mode and none when in normal mode. The rationale for this is one of consistency - "4: Consistency and Standards" usability heuristic. So, when I "Explore > My records" I only see those from the Mode that I am in, not a mixture with a flag to say which were entered in Training Mode.

As I commented [on Jun 30, 2023](https://github.com/BiologicalRecordsCentre/iRecord/issues/1294#issuecomment-1615101290), I still think that an action to delete/undo an import from the imports/my-list form would be a very valuable feature. This would support the "3: User Control and Freedom" usability heuristic in allowing a user to undo a mistake! Maybe this is in plan for the future?

**Comment by DavidHepper** [comment]
Yes, I'd support both @cubitmg 's suggestions. _Plus_ an explainer for 'The record cannot be found'.

**Comment by johnvanbreda** [comment]
For the v2 importer, I've added a training flag to the import metadata table in the develop branch, so we can track which imports are training ones. There is also a new automatic filter in the v2 imports list report so it will display the correct imports depending on your user profile setting.

**Once deployed, I'll need to copy the modifications to the v2 imports list report into the custom version used on iRecord so it respects the training flag.** 

@kitenetter remaining things to think about, as we now have 2 versions of the import list page. The first (https://irecord.org.uk/import/my) shows any import including v1 and v2, whereas the second (https://irecord.org.uk/imports/my-list) only shows imports from the v2 importer, but will be much faster to load and can show the metadata given for the import. 
1. Do we need a version for the old import method now, or can we just have the faster version that lists v2 imports only?
2. If we are keeping the old version of the list page which shows both v1 and v2 imports, should we add a link to the Explore page to allow you to view the records?
3. If so does it also need to support training mode (given that this may slow it down a bit)?
4. Does the new v2 import list need to show minimum and maximum record Ids?
5. Do we want to integrate new import-rollback features that Andy has been writing into the v2 import list so you get a link allowing a rollback?
6. For the list pages that show min and max record IDs for each import, do you want a message somewhere explaining they are not always contiguous?

**Comment by kitenetter** [comment]
@johnvanbreda some answers:

1. I suggest we implement for v2 only at the moment
2. NA
3. NA
4. I don't think we need to show the IDs - users can view records to see the IDs (and sort ascending or descending as required)
5. I think a rollback would be a very useful addition if it can be achieved
6. NA

**Comment by kitenetter** [comment]
The v2 listing has a column headed "Description", which I believe contains the metadata that can added in a text box during the import process. That text box is called "Import metadata" on the import page, so I think we should change the import list "Description" column to "Import metadata" for consistency.

**Comment by johnvanbreda** [comment]
I've made the report filter according to your account's training mode status. I've also renamed the Description column to Import metadata.

For the import reversal tool, I can add a link but having tested this functionality briefly I think it needs a bit of work to ensure it does not time out on a large import before wider use.

**Comment by johnvanbreda** [comment]
@kitenetter is there anything else we need to do to this page https://irecord.org.uk/imports/my-list or can we close this?

**Comment by sacrevert** [comment]
How does a user find these pages if they haven't been supplied with the URL?

**Comment by kitenetter** [comment]
@sacrevert a user will have to wait for me to find time to write the guidance and add it to an appropriate menu :)

**Comment by DavidHepper** [comment]
IMO it would be useful to have an entry in the **Imported into** column for "Training Mode data".

**Comment by sacrevert** [comment]
I was assuming that training mode imports were only shown in training mode and vice versa?

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by johnvanbreda** [comment]
@sacrevert that's correct - it's also consistent with the way other reporting pages work with training mode.

**Comment by cubitmg** [comment]
> [@sacrevert](https://github.com/sacrevert) that's correct - it's also consistent with the way other reporting pages work with training mode.

I think that is the right design, however, I don't think it is working. I have never (yet) made a production import, although I am about too soon. However, when in production (rather than training) mode I see several Training mode imports. See attached screenshot.  If I click on the View records link then no records are shown unless I switch to Training mode.

<img width="2254" height="1469" alt="Image" src="https://github.com/user-attachments/assets/bd9e5fcf-eac3-4470-9f85-4ce3de281c5f" />

**Comment by DavidHepper** [comment]
I'm also seeing Training Mode imports in normal mode. Example:
`Public data | 2024-10-15 11:02:21 | David Hepper | 1103 | Test of the importer process. Records may be deleted.`

and if I click **view records** the page shows 'No hits'
